modules/tensor_mechanics/src/materials/ADViscoplasticityStressUpdate.C

```diff
@@ -0,0 +1,387 @@
+    _creep_strain_old(getMaterialPropertyOld<RankTwoTensor>(_base_name + "creep_strain")),
+    _max_inelastic_increment(getParam<Real>("max_inelastic_increment")),
+    _power(getParam<Real>("power")),
+    _power_factor(_model == "LPS" ? (_power - 1.0) / (_power + 1.0) : 1.0),
+    _coefficient(getADMaterialProperty<Real>("coefficient")),
+    _gauge_stress(declareADProperty<Real>(_base_name + "gauge_stress")),
+    _intermediate_porosity(0.0),
+    _porosity_old(getMaterialPropertyOld<Real>(getParam<MaterialPropertyName>("porosity_name"))),
+    _verbose(getParam<bool>("verbose")),
+    _hydro_stress(0.0),
+    _identity_two(RankTwoTensor::initIdentity),
+    _dhydro_stress_dsigma(_identity_two / 3.0),
+    _derivative(0.0)
+{
+}
+
+template <ComputeStage compute_stage>
+MooseEnum
+ADViscoplasticityStressUpdate<compute_stage>::getModelEnum()
+{
+  return MooseEnum("LPS GTN", "LPS");
+}
+
+template <ComputeStage compute_stage>
+MooseEnum
+ADViscoplasticityStressUpdate<compute_stage>::getPoreShapeEnum()
+{
+  return MooseEnum("spherical cylindrical", "spherical");
+}
+
+template <ComputeStage compute_stage>
+void
+ADViscoplasticityStressUpdate<compute_stage>::initQpStatefulProperties()
+{
+  _effective_inelastic_strain[_qp] = 0.0;
+  _creep_strain[_qp].zero();
+}
+
+template <ComputeStage compute_stage>
+void
+ADViscoplasticityStressUpdate<compute_stage>::propagateQpStatefulProperties()
+{
+  _effective_inelastic_strain[_qp] = _effective_inelastic_strain_old[_qp];
+  _creep_strain[_qp] = _creep_strain_old[_qp];
+}
+
+template <ComputeStage compute_stage>
+void
+ADViscoplasticityStressUpdate<compute_stage>::updateState(
+    ADRankTwoTensor & elastic_strain_increment,
+    ADRankTwoTensor & inelastic_strain_increment,
+    const ADRankTwoTensor & /*rotation_increment*/,
+    ADRankTwoTensor & stress,
+    const RankTwoTensor & /*stress_old*/,
+    const ADRankFourTensor & elasticity_tensor,
+    const RankTwoTensor & elastic_strain_old)
+{
+  // Compute initial hydrostatic stress and porosity
+  if (_pore_shape == "cylindrical")
+    _hydro_stress = (stress(0, 0) + stress(1, 1)) / 2.0;
+  else
+    _hydro_stress = stress.trace() / 3.0;
+
+  // Subtract elastic strain from strain increment to find all inelastic strain increments
+  // calculated so far except the one that we're about to calculate
+  const ADRankTwoTensor inelastic_volumetric_increment =
+      _strain_increment[_qp] - elastic_strain_increment;
+  // Calculate intermdiate porosity from all inelastic strain increments calculated so far except
+  // the one that we're about to calculate
+  _intermediate_porosity =
+      (1.0 - _porosity_old[_qp]) * inelastic_volumetric_increment.trace() + _porosity_old[_qp];
+
+  // Compute intermediate equivalent stress
+  const ADRankTwoTensor dev_stress = stress.deviatoric();
+  const ADReal dev_stress_squared = dev_stress.doubleContraction(dev_stress);
+  const ADReal equiv_stress = MooseUtils::absoluteFuzzyEqual(dev_stress_squared, 0.0)
```

Comment:
What is this needed, by the way?  std::sqrt(1.5 * dev_stress_squared) gives zero if dev_stress_squared=0, right?  Or is it to guard against negative numbers?  I'm just concerned about 1E-16 being a huge stress for some people, while it's a tiny stress for other people.

Reply:
This is presumably to work around the fact that the derivative of sqrt is undefined at zero. If there is no guard here then MetaPhysicL will throw NaNs in the derivative vector and rightly so.

Reply:
Moreover, I am worried about the condition number for the user whose stress of 1e-16 is huge :-( is that really a thing? It seems like maybe a change in units would be in order in those circumstances?

Reply:
I would actually remove the fuzzy call though and simply compare directly to 0.0. This in its current form would actually set the expression to zero for any stress smaller than 1e-6.

Reply:
Comparing to 0 is probably the best way, good call. This was taken from ADRadialReturnStressUpdate techniques. I would be surprised if stresses of 1e-16 would be a thing, but good point it should be agnostic.

Reply:
I assume `!dev_stress_squared` will work for dual numbers here?

Reply:
No it won't work. That unary operator returns another `DualNumber`. You will just have to use `dev_stress_squared == 0.0`

Reply:
I'm not sure what derivative you're talking about.  I think that you really want d(equiv_stress)/d(stress_ij), or d(sqrt(equiv_stress))/d(stress_ij), both of which can be well defined as stress->0 (the latter is delta_ij as deviatoric->0).

Reply:
I agree that 1E-16 is probably a bit extreme.  But i think adding an Advanced user input that specifies "small stress" would be useful.

Reply:
> I'm not sure what derivative you're talking about.

I'm talking about the derivative of `sqrt(x)` or I guess more specifically here `sqrt(1.5*x)` which is the expression that will be evaluated here if the boolean expression evaulates to `false`. This is undefined at `x = 0`. AD unfortunately knows nothing the "bigger picture"; it only propagates through every elementary operation.